### PR TITLE
Fixup arm64 build; Remove amd64 hardcode from Makefile to make it more flexible;

### PR DESCRIPTION
**Pull request by brezerk**
HW: Rock64 (arm64v8)
OS: Gentoo

Hi

Thank you for nice DevOps tool. I was looking forward to build if for my arm64 k8s cluster and got some issues.

Here you are the changes I had to make to compile the binaries and build the images.

Also, i have noticed comments/discussions around #1761 , #1411 and #1474 so i have tried to update Makefile accordingly.

### CODE REVIEW

**Comment by 2opremio** [comment]
Thanks for your PR! Let me take a look.

File: Makefile [hunk 1]
```diff
@@ -58,15 +71,15 @@ build/.helm-operator.done: build/helm-operator build/kubectl build/helm docker/s
 
 build/fluxd: $(FLUXD_DEPS)
 build/fluxd: cmd/fluxd/*.go
-	CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o $@ $(LDFLAGS) -ldflags "-X main.version=$(shell ./docker/image-tag)" ./cmd/fluxd
+	CGO_ENABLED=0 GOOS=linux GOARCH=${GOARCH} go build -o $@ $(LDFLAGS) -ldflags "-X main.version=$(shell ./docker/image-tag)" ./cmd/fluxd
```
Comment (2opremio) [inline Makefile]: This breaks the development environment of some people (including me).

For instance, I develop in macOS (darwin) but I run flux in a local k8s cluster (amd64).

Comment (2opremio) [reply Makefile]: In fact, supporting fluxd in architectures other than amd64 would require also producing docker images. Ideally multiarchitecture images.

**Comment by 2opremio** [comment]
I am fine with a change which lets you compile for `arm64` but that it doesn't disrupt the current development environments. As it is, this PR doesn't comply with that.

For instance, you could set an environment variable indicating the target architecture when running `make`. Without the env variable it compiles as usual.

File: Makefile (hunk 1 above)
Comment (brezerk) [reply Makefile]: hi @2opremio 

Thank you for you comments.

> For instance, I develop in macOS (darwin) but I run flux in a local k8s cluster (amd64). The docker image produced (which uses build/fluxd) is expected to be amd64

I am fine with your suggestion regarding additional env variable to be passed to `make`, however, I am a bit confused.

Can you show what is GOARCH variable value on your macOS (darwin) please `go env GOARCH`? I would assume it should be `amd64`?

> In fact, supporting fluxd in architectures other than amd64 would require also producing docker images in that architecture. Ideally multi-architecture images.

I have build the arm64 images using current code. You are using alpine base images. Which seems to be multi-architecture images.

At least it seems to be working just fine without additional changes, see execution test:
```
bash-4.4$ docker run registry.internal-system.svc.cluster.local:5000/flux/flux:master-6e34d822 --version
master-6e34d822
bash-4.4$ docker run registry.internal-system.svc.cluster.local:5000/flux/helm-operator:master-6e34d822 --version
master-6e34d822
bash-4.4$ uname -a
Linux binging-hyena-jenkins-slave-7cdbbf7fb7-5mzhf 4.19.0-gentoo #5 SMP Sun Oct 28 03:36:38 EEST 2018 aarch64 Linux
```

Full build log:
```
root@97f7c6a73110:/go/src/github.com/weaveworks/flux# make
go install ./cmd/fluxctl
go install ./cmd/fluxd
CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -o build/fluxd  -ldflags "-X main.version=master-6e34d822" ./cmd/fluxd
mkdir -p cache/linux-arm64
curl -L -o cache/linux-arm64/kubectl-v1.11.3.tar.gz "https://dl.k8s.io/v1.11.3/kubernetes-client-linux-arm64.tar.gz"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   161  100   161    0     0    195      0 --:--:-- --:--:-- --:--:--   195
100 12.4M  100 12.4M    0     0  1561k      0  0:00:08  0:00:08 --:--:-- 1752k
[ linux-arm64 != "linux-amd64" ] || echo "14a70ac05c00fcfd7d632fc9e7a5fbc6615ce1b370bb1a0e506a24972d461493  cache/linux-arm64/kubectl-v1.11.3.tar.gz" | shasum -a 256 -c
tar -m --strip-components 3 -C ./cache/linux-arm64 -xzf cache/linux-arm64/kubectl-v1.11.3.tar.gz kubernetes/client/bin/kubectl
mv ./cache/linux-arm64/kubectl cache/linux-arm64/kubectl-v1.11.3
mkdir -p build
cp cache/linux-arm64/kubectl-v1.11.3 build/kubectl
if [ `basename build/kubectl` = "build" -a linux-arm64 = "linux-amd64" ]; then strip build/kubectl; fi
chmod a+x build/kubectl
mkdir -p ./build/docker/flux
cp docker/Dockerfile.flux build/fluxd build/kubectl docker/ssh_config docker/kubeconfig docker/verify_known_hosts.sh ./build/docker/flux/
docker build -t quay.io/weaveworks/flux -t quay.io/weaveworks/flux:master-6e34d822 \
        --build-arg VCS_REF="6e34d822605a9b616db9bda917c60a9381ebffaa" \
        --build-arg BUILD_DATE="2019-04-15T21:04:37Z" \
        -f build/docker/flux/Dockerfile.flux ./build/docker/flux
Sending build context to Docker daemon  107.3MB
Step 1/17 : FROM alpine:3.9
Get https://registry-1.docker.io/v2/: net/http: TLS handshake timeout
Makefile:61: recipe for target 'build/.flux.done' failed
make: *** [build/.flux.done] Error 1
root@97f7c6a73110:/go/src/github.com/weaveworks/flux# make
mkdir -p ./build/docker/flux
cp docker/Dockerfile.flux build/fluxd build/kubectl docker/ssh_config docker/kubeconfig docker/verify_known_hosts.sh ./build/docker/flux/
docker build -t quay.io/weaveworks/flux -t quay.io/weaveworks/flux:master-6e34d822 \
        --build-arg VCS_REF="6e34d822605a9b616db9bda917c60a9381ebffaa" \
        --build-arg BUILD_DATE="2019-04-15T22:17:33Z" \
        -f build/docker/flux/Dockerfile.flux ./build/docker/flux
Sending build context to Docker daemon  107.3MB
Step 1/17 : FROM alpine:3.9
3.9: Pulling from library/alpine
6f37394be673: Pull complete
Digest: sha256:28ef97b8686a0b5399129e9b763d5b7e5ff03576aa5580d6f4182a49c5fe1913
Status: Downloaded newer image for alpine:3.9
 ---> 660030ffd296
Step 2/17 : WORKDIR /home/flux
 ---> Running in 05b0d21257bb
Removing intermediate container 05b0d21257bb
 ---> 3032ec209046
Step 3/17 : RUN apk add --no-cache openssh ca-certificates tini 'git>=2.3.0' gnupg
 ---> Running in c8dbb77268a8
fetch http://dl-cdn.alpinelinux.org/alpine/v3.9/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.9/community/aarch64/APKINDEX.tar.gz
(1/38) Installing ca-certificates (20190108-r0)  
(2/38) Installing nghttp2-libs (1.35.1-r0)
(3/38) Installing libssh2 (1.8.2-r0)
(4/38) Installing libcurl (7.64.0-r1)
(5/38) Installing expat (2.2.6-r0)
(6/38) Installing pcre2 (10.32-r1)
(7/38) Installing git (2.20.1-r0)
(8/38) Installing libgpg-error (1.33-r0)
(9/38) Installing libassuan (2.5.1-r0)
(10/38) Installing libcap (2.26-r0)
(11/38) Installing ncurses-terminfo-base (6.1_p20190105-r0)
(12/38) Installing ncurses-terminfo (6.1_p20190105-r0)
(13/38) Installing ncurses-libs (6.1_p20190105-r0)
(14/38) Installing pinentry (1.1.0-r0)
Executing pinentry-1.1.0-r0.post-install
(15/38) Installing libbz2 (1.0.6-r6)
(16/38) Installing libgcrypt (1.8.4-r0)
(17/38) Installing gmp (6.1.2-r1)
(18/38) Installing nettle (3.4.1-r0)
(19/38) Installing libffi (3.2.1-r6)
(20/38) Installing p11-kit (0.23.14-r0)
(21/38) Installing libtasn1 (4.13-r0)
(22/38) Installing libunistring (0.9.10-r0)
(23/38) Installing gnutls (3.6.7-r0)
(24/38) Installing libksba (1.3.5-r0)
(25/38) Installing db (5.3.28-r1)
(26/38) Installing libsasl (2.1.27-r1)
(27/38) Installing libldap (2.4.47-r2)
(28/38) Installing npth (1.6-r0)
(29/38) Installing sqlite-libs (3.26.0-r3)
(30/38) Installing gnupg (2.2.12-r0)
(31/38) Installing openssh-keygen (7.9_p1-r4)
(32/38) Installing libedit (20181209.3.1-r0)
(33/38) Installing openssh-client (7.9_p1-r4)
(34/38) Installing openssh-sftp-server (7.9_p1-r4)
(35/38) Installing openssh-server-common (7.9_p1-r4)
(36/38) Installing openssh-server (7.9_p1-r4)
(37/38) Installing openssh (7.9_p1-r4)
(38/38) Installing tini (0.18.0-r0)
Executing busybox-1.29.3-r10.trigger
Executing ca-certificates-20190108-r0.trigger
OK: 47 MiB in 52 packages
Removing intermediate container c8dbb77268a8
 ---> 30e56ff059be
Step 4/17 : ADD ./verify_known_hosts.sh /home/flux/verify_known_hosts.sh
 ---> df6dd5a96383
Step 5/17 : RUN ssh-keyscan github.com gitlab.com bitbucket.org ssh.dev.azure.com vs-ssh.visualstudio.com >> /etc/ssh/ssh_known_hosts &&     sh /home/flux/verify_known_hosts.sh /etc/ssh/ssh_known_hosts &&     rm /home/flux/verify_known_hosts.sh
 ---> Running in 487305c11721
# github.com:22 SSH-2.0-babeld-3e010779
# github.com:22 SSH-2.0-babeld-3e010779
# github.com:22 SSH-2.0-babeld-3e010779
# gitlab.com:22 SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
# gitlab.com:22 SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
# gitlab.com:22 SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
# bitbucket.org:22 SSH-2.0-conker_1.1.57-7474e6d app-130
# bitbucket.org:22 SSH-2.0-conker_1.1.57-7474e6d app-131
# bitbucket.org:22 SSH-2.0-conker_1.1.57-7474e6d app-154
# ssh.dev.azure.com:22 SSH-2.0-SSHBlackbox.10
# ssh.dev.azure.com:22 SSH-2.0-SSHBlackbox.10
# ssh.dev.azure.com:22 SSH-2.0-SSHBlackbox.10
# vs-ssh.visualstudio.com:22 SSH-2.0-SSHBlackbox.10
# vs-ssh.visualstudio.com:22 SSH-2.0-SSHBlackbox.10
# vs-ssh.visualstudio.com:22 SSH-2.0-SSHBlackbox.10
Removing intermediate container 487305c11721
 ---> d513d0402167
Step 6/17 : COPY ./ssh_config /etc/ssh/ssh_config
 ---> 8f01db15160d
Step 7/17 : COPY ./kubectl /usr/local/bin/
 ---> cfe90d4dd312
Step 8/17 : LABEL maintainer="Weaveworks <help@weave.works>"       org.opencontainers.image.title="flux"       org.opencontainers.image.description="The Flux daemon, for synchronising your cluster with a git repo, and deploying new images"       org.opencontainers.image.url="https://github.com/weaveworks/flux"       org.opencontainers.image.source="git@github.com:weaveworks/flux"       org.opencontainers.image.vendor="Weaveworks"       org.label-schema.schema-version="1.0"       org.label-schema.name="flux"       org.label-schema.description="The Flux daemon, for synchronising your cluster with a git repo, and deploying new images"       org.label-schema.url="https://github.com/weaveworks/flux"       org.label-schema.vcs-url="git@github.com:weaveworks/flux"       org.label-schema.vendor="Weaveworks"
 ---> Running in 91994ecac6fd

Removing intermediate container 91994ecac6fd
 ---> 3e6b8cdf178d
Step 9/17 : ENTRYPOINT [ "/sbin/tini", "--", "fluxd" ]
 ---> Running in ba11317d2d0d
Removing intermediate container ba11317d2d0d
 ---> 44c88c6a6609
Step 10/17 : COPY --from=quay.io/squaremo/kubeyaml:0.5.1 /usr/lib/kubeyaml /usr/lib/kubeyaml/
0.5.1: Pulling from squaremo/kubeyaml
3489d1c4660e: Pull complete
e9ae08d466d1: Pull complete
Digest: sha256:9cc50e2f36fcae6da0f42a0e186c8d38fa0bcd53329abacda0fe1d2b4fa2d9d5
Status: Downloaded newer image for quay.io/squaremo/kubeyaml:0.5.1
 ---> 327a97c0a950
Step 11/17 : ENV PATH=/bin:/usr/bin:/usr/local/bin:/usr/lib/kubeyaml
 ---> Running in 4aa66e998130

Removing intermediate container 4aa66e998130
 ---> 76df5486242c
Step 12/17 : RUN [ ! -e /etc/nsswitch.conf ] && echo 'hosts: files dns' > /etc/nsswitch.conf
 ---> Running in f48f99bea98c
Removing intermediate container f48f99bea98c
 ---> 23558c0b3ea0
Step 13/17 : COPY ./kubeconfig /root/.kube/config
 ---> e52c73243043
Step 14/17 : COPY ./fluxd /usr/local/bin/
 ---> e83779650ebd
Step 15/17 : ARG BUILD_DATE
 ---> Running in 5059e452c7e4

Removing intermediate container 5059e452c7e4
 ---> a1981a5a0690
Step 16/17 : ARG VCS_REF
 ---> Running in 065c4869d2ce
Removing intermediate container 065c4869d2ce
 ---> a1581ec7339c
Step 17/17 : LABEL org.opencontainers.image.revision="$VCS_REF"       org.opencontainers.image.created="$BUILD_DATE"       org.label-schema.vcs-ref="$VCS_REF"       org.label-schema.build-date="$BUILD_DATE"
 ---> Running in f016a58a5625
Removing intermediate container f016a58a5625
 ---> 362bb65f0a69
Successfully built 362bb65f0a69
Successfully tagged quay.io/weaveworks/flux:latest
Successfully tagged quay.io/weaveworks/flux:master-6e34d822
touch build/.flux.done
CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -o build/helm-operator  -ldflags "-X main.version=master-6e34d822" ./cmd/helm-operator
mkdir -p cache/linux-arm64
curl -L -o cache/linux-arm64/helm-2.13.0.tar.gz "https://storage.googleapis.com/kubernetes-helm/helm-v2.13.0-linux-arm64.tar.gz"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 19.7M  100 19.7M    0     0  7236k      0  0:00:02  0:00:02 --:--:-- 7236k
[ linux-arm64 != "linux-amd64" ] || echo "15eca6ad225a8279de80c7ced42305e24bc5ac60bb7d96f2d2fa4af86e02c794  cache/linux-arm64/helm-2.13.0.tar.gz" | shasum -a 256 -c
tar -m -C ./cache -xzf cache/linux-arm64/helm-2.13.0.tar.gz linux-arm64/helm
mv cache/linux-arm64/helm cache/linux-arm64/helm-2.13.0
mkdir -p build
cp cache/linux-arm64/helm-2.13.0 build/helm
if [ `basename build/helm` = "build" -a linux-arm64 = "linux-amd64" ]; then strip build/helm; fi
chmod a+x build/helm
mkdir -p ./build/docker/helm-operator
cp docker/Dockerfile.helm-operator build/helm-operator build/kubectl build/helm docker/ssh_config docker/verify_known_hosts.sh docker/helm-repositories.yaml ./build/docker/helm-operator/
docker build -t quay.io/weaveworks/helm-operator -t quay.io/weaveworks/helm-operator:master-6e34d822 \
        --build-arg VCS_REF="6e34d822605a9b616db9bda917c60a9381ebffaa" \
        --build-arg BUILD_DATE="2019-04-15T22:17:33Z" \
        -f build/docker/helm-operator/Dockerfile.helm-operator ./build/docker/helm-operator
Sending build context to Docker daemon  138.7MB  
Step 1/17 : FROM alpine:3.9
 ---> 660030ffd296
Step 2/17 : WORKDIR /home/flux
 ---> Using cache
 ---> 3032ec209046
Step 3/17 : RUN apk add --no-cache openssh ca-certificates tini 'git>=2.3.0'
 ---> Running in 64089cb3d6cb
fetch http://dl-cdn.alpinelinux.org/alpine/v3.9/main/aarch64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.9/community/aarch64/APKINDEX.tar.gz
(1/18) Installing ca-certificates (20190108-r0)  
(2/18) Installing nghttp2-libs (1.35.1-r0)
(3/18) Installing libssh2 (1.8.2-r0)
(4/18) Installing libcurl (7.64.0-r1)
(5/18) Installing expat (2.2.6-r0)
(6/18) Installing pcre2 (10.32-r1)
(7/18) Installing git (2.20.1-r0)
(8/18) Installing openssh-keygen (7.9_p1-r4)
(9/18) Installing ncurses-terminfo-base (6.1_p20190105-r0)
(10/18) Installing ncurses-terminfo (6.1_p20190105-r0)
(11/18) Installing ncurses-libs (6.1_p20190105-r0)
(12/18) Installing libedit (20181209.3.1-r0)
(13/18) Installing openssh-client (7.9_p1-r4)
(14/18) Installing openssh-sftp-server (7.9_p1-r4)
(15/18) Installing openssh-server-common (7.9_p1-r4)
(16/18) Installing openssh-server (7.9_p1-r4)
(17/18) Installing openssh (7.9_p1-r4)
(18/18) Installing tini (0.18.0-r0)
Executing busybox-1.29.3-r10.trigger
Executing ca-certificates-20190108-r0.trigger
OK: 32 MiB in 32 packages
Removing intermediate container 64089cb3d6cb
 ---> aba040d898ab
Step 4/17 : ADD ./verify_known_hosts.sh /home/flux/verify_known_hosts.sh
 ---> 77438b498c06
Step 5/17 : RUN ssh-keyscan github.com gitlab.com bitbucket.org ssh.dev.azure.com vs-ssh.visualstudio.com >> /etc/ssh/ssh_known_hosts &&     sh /home/flux/verify_known_hosts.sh /etc/ssh/ssh_known_hosts &&     rm /home/flux/verify_known_hosts.sh
 ---> Running in c6db3e5d2ca8
# github.com:22 SSH-2.0-babeld-3e010779
# github.com:22 SSH-2.0-babeld-3e010779
# github.com:22 SSH-2.0-babeld-3e010779
# gitlab.com:22 SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
# gitlab.com:22 SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
# gitlab.com:22 SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
# bitbucket.org:22 SSH-2.0-conker_1.1.57-7474e6d app-155
# bitbucket.org:22 SSH-2.0-conker_1.1.57-7474e6d app-191
# bitbucket.org:22 SSH-2.0-conker_1.1.57-7474e6d app-131
# ssh.dev.azure.com:22 SSH-2.0-SSHBlackbox.10
# ssh.dev.azure.com:22 SSH-2.0-SSHBlackbox.10
# ssh.dev.azure.com:22 SSH-2.0-SSHBlackbox.10
# vs-ssh.visualstudio.com:22 SSH-2.0-SSHBlackbox.10
# vs-ssh.visualstudio.com:22 SSH-2.0-SSHBlackbox.10
# vs-ssh.visualstudio.com:22 SSH-2.0-SSHBlackbox.10
Removing intermediate container c6db3e5d2ca8
 ---> 01cedeb6cb95
Step 6/17 : COPY ./ssh_config /etc/ssh/ssh_config
 ---> 859a5cbb92d0
Step 7/17 : COPY ./kubectl /usr/local/bin/
 ---> a842653994a4
Step 8/17 : COPY ./helm /usr/local/bin/
 ---> fba0517b47aa
Step 9/17 : LABEL maintainer="Weaveworks <help@weave.works>"       org.opencontainers.image.title="flux-helm-operator"       org.opencontainers.image.description="The Flux Helm operator, for releasing Helm charts according to git"       org.opencontainers.image.url="https://github.com/weaveworks/flux"       org.opencontainers.image.source="git@github.com:weaveworks/flux"       org.opencontainers.image.vendor="Weaveworks"       org.label-schema.schema-version="1.0"       org.label-schema.name="flux-helm-operator"       org.label-schema.description="The Flux Helm operator, for releasing Helm charts according to git"       org.label-schema.url="https://github.com/weaveworks/flux"       org.label-schema.vcs-url="git@github.com:weaveworks/flux"       org.label-schema.vendor="Weaveworks"
 ---> Running in f8fd436e2d9e
Removing intermediate container f8fd436e2d9e
 ---> 3c9f7f547293
Step 10/17 : ENTRYPOINT [ "/sbin/tini", "--", "helm-operator" ]
 ---> Running in 022664835f5e
Removing intermediate container 022664835f5e
 ---> e8ecd8f3b094
Step 11/17 : ENV HELM_HOME=/var/fluxd/helm
 ---> Running in 3e08f184f7d1
Removing intermediate container 3e08f184f7d1
 ---> f06519789db4
Step 12/17 : COPY ./helm-repositories.yaml /var/fluxd/helm/repository/repositories.yaml
 ---> 1baab39b8250
Step 13/17 : RUN mkdir -p /var/fluxd/helm/repository/cache/
 ---> Running in 7c78af2402d8
Removing intermediate container 7c78af2402d8
 ---> 77dfa67be759
Step 14/17 : COPY ./helm-operator /usr/local/bin/
 ---> 0499d9e88a1a
Step 15/17 : ARG BUILD_DATE
 ---> Running in d8cfc4b431b5
Removing intermediate container d8cfc4b431b5
 ---> 4c2427c1d2ff
Step 16/17 : ARG VCS_REF
 ---> Running in 247fee51b821
Removing intermediate container 247fee51b821
 ---> c1b922f05d36
Step 17/17 : LABEL org.opencontainers.image.revision="$VCS_REF"       org.opencontainers.image.created="$BUILD_DATE"       org.label-schema.vcs-ref="$VCS_REF"       org.label-schema.build-date="$BUILD_DATE"
 ---> Running in 5b670c189b26
Removing intermediate container 5b670c189b26
 ---> 2817bce398b3
Successfully built 2817bce398b3
Successfully tagged quay.io/weaveworks/helm-operator:latest
Successfully tagged quay.io/weaveworks/helm-operator:master-6e34d822
touch build/.helm-operator.done
```

**Comment by brezerk** [comment]
Hi @2opremio ,

I have updated PR according to the comments: new ARCH variable has been added. It has default value set to `adm64`. `make ARCH=<arch>` could be used to choose the desired target arch.

>You should also choose the correct base docker image (matching the target architecture).

The base alpine images are multi-arch ones by default. So no need to change anything here. I have tested the generated images and they looks pretty fine for me:

```
[ marauder ] brezerk@pts/3:77  ~/develop/helm/flux $
 04/16/19 12:47:18 EEST > kubectl -n flux get pods -o wide
NAME                                  READY   STATUS    RESTARTS   AGE    IP             NODE         NOMINATED NODE   READINESS GATES
flux-6d694c876d-bnhws                 1/1     Running   0          126m   10.244.1.225   cerberus02   <none>           <none>
flux-helm-operator-6db4677799-d4wg8   1/1     Running   0          126m   10.244.2.211   cerberus01   <none>           <none>
flux-memcached-757756884-7czts        1/1     Running   0          142m   10.244.4.132   brentoo01    <none>           <none>
You have new mail.
[ marauder ] brezerk@pts/3:78  ~/develop/helm/flux $
 04/16/19 14:37:27 EEST > kubectl get nodes -l beta.kubernetes.io/arch=arm64 
NAME         STATUS   ROLES    AGE    VERSION
cerberus01   Ready    <none>   169d   v1.13.4
cerberus02   Ready    <none>   21d    v1.13.4
marauder     Ready    master   170d   v1.13.4
[ marauder ] brezerk@pts/3:79  ~/develop/helm/flux $
 04/16/19 14:37:37 EEST > 
```

Note: if at some point you would like to build official arm/arm64 images you may need to adopt `binfmt_misc` approach as your current circle docker executor doesn't natively support building ARM images. See https://github.com/prometheus/prometheus/pull/5031 for details.

Thanks

File: Makefile (hunk 1 above)
Comment (2opremio) [reply Makefile]: >> This breaks the development environment of some people (including me).
> I am fine with your suggestion regarding additional env variable to be passed to make, however, I am a bit confused.
>
> Can you show what is GOARCH variable value on your macOS (darwin) please go env GOARCH? I would assume it should be amd64?

You are right, it's amd64, so it wouldn't break my dev environment.

>  You are using alpine base images. Which seems to be multi-architecture images.

Ahá, perfect then!

**Comment by 2opremio** [comment]
Can you update `bin/upload-binaries` to include arm?

Also, optionally:
1. For extra points, could you make sure that the arm64 binaries of `kubectl` and `helm` are checksummed and verified before being put into the docker image?
2. I am happy with the `ARCH` env approach. I would be even happier if we produced a `fluxd` multiarch docker image that we could release. The multi-arch docker build should be optional and only enabled for releases since producing a multiarch docker image just slows down normal development.

(1) doesn't make much sense without (2). And, as mentioned, they are both optional but desirable.

**Comment by brezerk** [comment]
Sure. God points. Will take a look.

**Comment by 2opremio** [comment]
Great, can you also squash the commits once you are done?

**Comment by brezerk** [comment]
Sure. One question tho...

```
if [ `basename $@` = "build" -a $(CURRENT_OS_ARCH) = "linux-amd64" ]; then strip $@; fi
[ $* != "linux-amd64" ] || echo "$(KUBECTL_CHECKSUM)  cache/$*/kubectl-$(KUBECTL_VERSION).tar.gz" | shasum -a 256 -c
```

is it b/c of windows and macos systems used for development purposes do not have strip/shasum binaries? will it be ok if i just change it to `$GOOS = "linux"` instead?

**Comment by 2opremio** [comment]
> ```
> if [ `basename $@` = "build" -a $(CURRENT_OS_ARCH) = "linux-amd64" ]; then strip $@; fi
> [ $* != "linux-amd64" ] || echo "$(KUBECTL_CHECKSUM)  cache/$*/kubectl-$(KUBECTL_VERSION).tar.gz" | shasum -a 256 -c
> ```
> is it b/c of windows and macos systems used for development purposes do not have strip/shasum binaries? will it be ok if i just change it to `$GOOS = "linux"` instead?

Only partly. It's because `strip` won't match the architecure/OS if the binary being built. With your current change I think it should be  `[ $* != "${ARCH}-linux"  ] ...`. But you may need to change this when doing multiarch docker images.

**Comment by brezerk** [comment]
hm...

```
component=client url=https://quay.io/v2/ status="401 UNAUTHORIZED"
component=client repo=quay.io/weaveworks/kured auth="<zero creds>" api=https://quay.io/v2/
component=client url="https://quay.io/v2/auth?scope=repository%3Aweaveworks%2Fkured%3Apull&service=quay.io" status="400 BAD REQUEST"
component=warmer canonical_name=quay.io/weaveworks/kured auth={map[]} err="requesting tags: Get https://quay.io/v2/weaveworks/kured/tags/list: unknown: Namespace weaveworks has been disabled. Please contact a system administrator."
component=warmer stopping=true
--- FAIL: TestWarming_WarmerWriteCacheRead (10.00s)
```

does not seems to be an issue related to my changes tho....

**Comment by 2opremio** [comment]
No, that's #1948 :S

**Comment by brezerk** [comment]
Mkay. After merge with master tests are passed. But now it looks like I squashed the commits in an odd way so PR looks... weird? :)

**Comment by hiddeco** [comment]
@brezerk it would be better to rebase on `origin/master` (or probably `upstream/master` in your case) instead of merging `master` into your branch, this way you omit a merge commit. Also, the commit message is a bit long, `<=50` characters is better.

We just made some changes to the tests to pull images from Docker Hub instead of Quay.io due to it being unavailable to us. So an additional rebase is necessary.

**Comment by brezerk** [comment]
git rebase... ok will take a look, thanks. I am not git wizard yet, sorry for all this spam >_<

**Comment by 2opremio** [comment]
@brezerk did you close the PR intentionally?

**Comment by brezerk** [comment]
@2opremio Well... I had to reset master to upstream/master and PR got closed automatically sine there was no changes. I have created a new one #1950

**Pull request closed without merging**